test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -123,6 +123,227 @@ void testExprSimple02() {
   }
 }
 
+void assertForRange(For* f, int expected_start, int expected_stop) {
+  ASSERT_NE(f, nullptr);
+  const IntImm* start = dynamic_cast<const IntImm*>(f->start());
+  ASSERT_NE(start, nullptr);
+  ASSERT_EQ(start->value(), expected_start);
+  const IntImm* stop = dynamic_cast<const IntImm*>(f->stop());
+  ASSERT_NE(stop, nullptr);
+  ASSERT_EQ(stop->value(), expected_stop);
+}
+
+void testExprSliceHeadWhenFactorEqualsSize() {
+  // When factor equals the For loop's original size, keep using the original
+  // For loop.
+  KernelScope kernel_scope;
+  auto func = [](const ExprHandle& x) {
+    return ExprHandle(1.0f) + cast<float>(x);
+  };
+  Tensor* tensor = Compute("f", {{10, "x"}}, func);
+  LoopNest l({tensor});
+  For* head;
+  For* tail;
+  std::vector<For*> loops = l.getLoopStmtsFor(tensor);
+  l.sliceHead(loops[0], 10, &head, &tail);
+
+  ASSERT_EQ(head, loops[0]);
+  ASSERT_EQ(tail, nullptr);
+
+  Stmt* stmt = l.root_stmt();
+  Stmt* simplified = IRSimplifier::simplify(stmt);
+  Block* body = dynamic_cast<Block*>(simplified);
+  ASSERT_EQ(body->nstmts(), 1);
+
+  For* loop = dynamic_cast<For*>(*body->begin());
+  assertForRange(loop, 0, 10);
+}
+
+void testExprSliceHead() {
+  KernelScope kernel_scope;
+  auto func = [](const ExprHandle& x) {
+    return ExprHandle(1.0f) + cast<float>(x);
+  };
+  Tensor* tensor = Compute("f", {{10, "x"}}, func);
+  LoopNest l({tensor});
+  For* head;
+  For* tail;
+  std::vector<For*> loops = l.getLoopStmtsFor(tensor);
+  l.setGPUBlockIndex(loops[0], LoopOptions::IDX_X);
+  l.sliceHead(loops[0], 4, &head, &tail);
+
+  ASSERT_NE(head, nullptr);
+  ASSERT_NE(head, loops[0]);
+  ASSERT_NE(tail, nullptr);
+  ASSERT_NE(tail, loops[0]);
+
+  ASSERT_TRUE(head->loop_options().is_gpu_block_index());
+  ASSERT_EQ(head->loop_options().gpu_block_index(), LoopOptions::IDX_X);
+  ASSERT_TRUE(tail->loop_options().is_gpu_block_index());
+  ASSERT_EQ(tail->loop_options().gpu_block_index(), LoopOptions::IDX_X);
```

Comment:
This is interesting (good), we should test that this actually works in Cuda before landing this change. e.g. your test creates this:

```
for x in 0..4: \\ blockIdx.x
  F[x] = 1 + x;
for x in 4..10: \\ blockIdx.x
  F[x] = 1 + x;
```
But this IR will break the CudaCodgen which needs the loop options for Blocks and Thread dimensions to start at zero. We'll need to normalize these (using the existing normalize transform perhaps) - I'm not sure if we want to normalize by default, or make the user do it, or add a bool to the API to normalize.

Assuming that is fixed we'll get this when we do metavar replacement in the Cuda Backend it gets turned into this:
```
F[blockIdx.x] = 1 + blockIdx.x;
F[blockIdx.x+4] = 1 + blockIdx.x + 4;
```
(where the Block dims are (6))

This will do too much work (12 writes instead of 10) unless we mask the first store with a Cond statement. The CudaCodgen should be doing that but it's not well tested.

Can you write a test like this (maybe exactly like this?) in test_cuda.cpp and make sure that a) the masking works, and b) it produces the correct result. Recommend pulling master, since I landed a change to metavar rewriting in the Cuda Codgen this morning.

Reply:
In `splitWithTail`, the outer loop inherits loop options, I'm wondering why the tail loop doesn't inherit the options?

Reply:
My intention was to test the loop options are kept after slicing.
Didn't think that far about CudaCodegen.
What about keeping each transformation independent, and letting the user do normalization after slicing if the gpu related options are set?

Reply:
I'm fine generally with that approach, but for it to work each transform must leave the IR in a valid state. This transform does not since the second loop is not normalized but does potentially have an axis binding. There are two approaches here: you can fix the second loop so that it works correctly with the axis binding, or you can not propagate the loop options to the second loop. The second is easier but probably doesn't do what the caller expects, which may cause issues down the line - if you choose this route please be explicit in the comments.

SplitWithTail is kind of an interesting case because you basically never want it on the GPU (you want SplitWithMask instead), so we do make a decision about what to do with LoopOptions but performance isn't really a consideration.

Reply:
In this PR, I'll leave LoopOptions unpropagated in sliceHead and sliceTail.
I'll try to fix this in a future PR.

Reply:
I don't think that's a great solution either - we'll silently drop previously set loop options which is definitely not what the caller wants.

I think the right compromise here in the short term is to propagate the loop options to the unsliced loop only, e.g. the second loop in SliceHead and the first in SliceTail. This is analogous to what we do in SplitWithTail where we don't propagate any options to the tail. Unlike splitWithTail we could handle it better than this with Slice, but I think it's fine not to in this PR.